form/data_mappers.rst

```diff
@@ -0,0 +1,197 @@
+    using a data mapper, you should configure the ``empty_data`` with a closure
+    as described in
+    :ref:`How to Configure empty Data for a Form Class <forms-empty-data-closure>`.
+
+The red, green and blue form fields have to be mapped to the constructor
+arguments and the ``Color`` instance has to be mapped to red, green and blue
+form fields. Recognize a familair pattern? It's time for a data mapper!
+
+.. code-block:: php
+
+    // src/AppBundle/Form/DataMapper/ColorMapper.php
+    namespace AppBundle\Form\DataMapper;
+
+    use AppBundle\Colors\Color;
+    use Symfony\Component\Form\DataMapperInterface;
+    use Symfony\Component\Form\Exception\UnexpectedTypeException;
+
+    class ColorMapper extends DataMapperInterface
+    {
+        public function mapDataToForms($data, $forms)
+        {
+            // there is no data yet, a new color will be created
+            if (null === $data) {
+                return;
+            }
+
+            // invalid data type, this message will not be shown to the user (see below)
+            if (!$data instanceof Color) {
+                throw new UnexpectedTypeException($data, Color::class);
+            }
+
+            $forms = iterator_to_array($forms);
+
+            // set form field values
+            $forms['red']->setData($data->getRed());
+            $forms['green']->setData($data->getGreen());
+            $forms['blue']->setData($data->getBlue());
+        }
+
+        public function mapFormsToData($forms, &$data)
+        {
+            $forms = iterator_to_array($forms);
+
+            // get form field values
+            $red = $forms['red']->getData();
+            $green = $forms['green']->getData();
+            $blue = $forms['blue']->getData();
+
+            $data = new Color($red, $green, $blue);
+        }
+    }
+
+All data mappers have to implement :class:`Symfony\\Component\\Form\\DataMapperInterface`.
+This interface contains two methods:
+
+``mapDataToForms()``
+    Recieves the current view data and a :phpclass:`RecursiveIteratorIterator`
+    with all form fields. It's task is to map the view data to the form fields.
+    The view data might be ``null`` if there is no data yet (e.g. in a "Create
+    new color" form);
+
+``mapFormsToData($forms, &$data)``
+    Maps the form field values to the ``$data`` property, which is changed by
+    reference. It recieves the form fields in a :phpclass:`RecursiveIteratorIterator`
+    and the view data.
+
+When an errors occurs, a
+:class:`Symfony\\Component\\Form\\Exception\\TransformationFailedException` or
+one of its subclasses should be thrown. The exception message won't be shown to
+the user. Instead, the ``invalid_message`` option will be used.
+
+.. caution::
+
+    The data passed to the mapper is *not yet validated*. This means that your
+    objects should allow being created in an invalid state in order to produce
+    user-friendly errors in the form.
```

Comment:
If the underlying fields use proper form types (like an `IntegerType` for an object property typehinted `int` in the constructor), the field itself will throw a `TransformationFailedException` which will be rendered as an error on the view. Important part is that the field won't be mapped back and (IIRC) keeps original value. Thus, in most cases, I think it won't cause any issue for scalar typehints.

See https://github.com/ogizanagi/symfony/commit/4f76c21f8109310c6938f7f6fc31d16d9b6ec074 and [FormTypeTest.php#L745-L752](https://github.com/ogizanagi/symfony/blob/4f76c21f8109310c6938f7f6fc31d16d9b6ec074/src/Symfony/Component/Form/Tests/Extension/Core/Type/FormTypeTest.php#L745-L752)
(it doesn't use typehint, but a `\DomainException` is thrown if `Money::amount` isn't numeric. The case is similar).

Appart from the above point, catching the error and throwing a `TranformationFailedException` ourself is a good safe guard (I recommend to use the `invalid_message` option in order to render a less cryptic error message, but that's limited).